pkg/template/nstemplatetiers/nstemplatetier_generator.go

```diff
@@ -0,0 +1,466 @@
+	tiers := make([]string, 0, len(t.templatesByTier))
+	for tier := range t.templatesByTier {
+		tiers = append(tiers, tier)
+	}
+	sort.Strings(tiers)
+	for _, tier := range tiers {
+		tierData := t.templatesByTier[tier]
+		basedOnTierFileRevision := ""
+		var parameters []templatev1.Parameter
+		if tierData.basedOnTier != nil {
+			parameters = tierData.basedOnTier.Parameters
+			basedOnTierFileRevision = tierData.rawTemplates.basedOnTier.revision
+			tierData = t.templatesByTier[tierData.basedOnTier.From]
+		}
+		tierTemplates, err := t.newTierTemplates(basedOnTierFileRevision, tierData, tier, parameters)
+		if err != nil {
+			return err
+		}
+		t.templatesByTier[tier].tierTemplates = tierTemplates
+	}
+
+	return nil
+}
+
+func (t *TierGenerator) newTierTemplates(basedOnTierFileRevision string, tierData *tierData, tier string, parameters []templatev1.Parameter) ([]*toolchainv1alpha1.TierTemplate, error) {
+	decoder := serializer.NewCodecFactory(t.scheme).UniversalDeserializer()
+
+	// namespace templates
+	kinds := make([]string, 0, len(tierData.rawTemplates.namespaceTemplates))
+	for kind := range tierData.rawTemplates.namespaceTemplates {
+		kinds = append(kinds, kind)
+	}
+	tierTmpls := []*toolchainv1alpha1.TierTemplate{}
+	sort.Strings(kinds)
+	for _, kind := range kinds {
+		tmpl := tierData.rawTemplates.namespaceTemplates[kind]
+		tierTmpl, err := t.newTierTemplate(decoder, basedOnTierFileRevision, tier, kind, tmpl, parameters)
+		if err != nil {
+			return nil, err
+		}
+		tierTmpls = append(tierTmpls, tierTmpl)
+	}
+	// space roles templates
+	roles := make([]string, 0, len(tierData.rawTemplates.spaceroleTemplates))
+	for role := range tierData.rawTemplates.spaceroleTemplates {
+		roles = append(roles, role)
+	}
+	sort.Strings(roles)
+	for _, role := range roles {
+		tmpl := tierData.rawTemplates.spaceroleTemplates[role]
+		tierTmpl, err := t.newTierTemplate(decoder, basedOnTierFileRevision, tier, role, tmpl, parameters)
+		if err != nil {
+			return nil, err
+		}
+		tierTmpls = append(tierTmpls, tierTmpl)
+	}
+	// cluster resources templates
+	if tierData.rawTemplates.clusterTemplate != nil {
+		tierTmpl, err := t.newTierTemplate(decoder, basedOnTierFileRevision, tier, toolchainv1alpha1.ClusterResourcesTemplateType, *tierData.rawTemplates.clusterTemplate, parameters)
+		if err != nil {
+			return nil, err
+		}
+		tierTmpls = append(tierTmpls, tierTmpl)
+	}
+	return tierTmpls, nil
+}
+
+// createTierTemplates creates all TierTemplate resources from the tier map
+func (t *TierGenerator) createTierTemplates() error {
+	// create the templates
+	for _, tierTmpls := range t.templatesByTier {
+		for _, tierTmpl := range tierTmpls.tierTemplates {
+			log.Info("creating TierTemplate", "namespace", tierTmpl.Namespace, "name", tierTmpl.Name)
+			// using the "standard" client since we don't need to support updates on such resources, they should be immutable
+			if _, err := t.ensureObject(tierTmpl, false); err != nil {
```

Comment:
the second parameter wasn't immediately clear to me, but I like the new `ensureObject` function.